src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,115 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizer;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.CostModel;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleInstances;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	protected final StoppingConditionForIterativeImprovement condition;
+	protected final Random rng;
+
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt, final StoppingConditionForIterativeImprovement x) {
+		assert ctxt != null;
+		assert x != null;
+		context = ctxt;
+		condition = x;
+		rng = new Random();
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+		return optimize( context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false) );
+	}
+
+	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
+		// The best plan we have found so far. As we have only found one plan, it is the best one so far.
+		PhysicalPlan bestPlan = initialPlan;
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel(); //Created as a value and stored rather than just using the getCostModel, because the same model is used later too.

```

Comment:
Gotcha. I'll assume the cost in performance for this would be negligible. I know it always gives me the same CostModel, but I've been taught that each extra call like this is more expensive computing-wise than simply storing a variable once and then re-using it.

Reply:
In principle you are right. However, for an engine like ours, the difference would not be measurable I guess. In fact, I would even assume that the Java compiler can even figure out that the access to the underlying CostModel object can be inlined.